apps/extension/src/offscreen.ts

```diff
@@ -0,0 +1,80 @@
+import type { JsonValue } from '@bufbuild/protobuf';
+import { OffscreenRequest } from '@penumbra-zone/types/src/internal-msg/offscreen-types';
+
+export const isOffscreenRequest = (req: unknown): req is OffscreenRequest =>
+  req != null &&
+  typeof req === 'object' &&
+  'type' in req &&
+  typeof req.type === 'string' &&
+  req.type === 'BUILD_ACTION';
+
+export const offscreenMessageHandler = (
+  req: unknown,
+  _: chrome.runtime.MessageSender,
+  sendResponse: (x: JsonValue) => void,
+) => {
+  if (!isOffscreenRequest(req)) return false;
+
+  buildActionHandler(req, sendResponse);
+  return true;
+};
+
+chrome.runtime.onMessage.addListener(offscreenMessageHandler);
+
+export const buildActionHandler = (
+  jsonReq: OffscreenRequest,
+  responder: (r: JsonValue) => void,
+) => {
+  // Destructure the data object to get individual fields
+  const { transactionPlan, witness, fullViewingKey, length } = jsonReq.request;
+
+  // Spawn web workers
+  const workerPromises: Promise<JsonValue>[] = [];
+  for (let i = 0; i < length; i++) {
+    workerPromises.push(spawnWorker(transactionPlan, witness, fullViewingKey, i));
+  }
+
+  // Wait for promises to resolve and construct response format
+  void Promise.all(workerPromises).then(batchActions => {
+    responder({
+      type: 'BUILD_ACTION',
+      data: batchActions,
+    });
```

Comment:
Can you type this with: 

```typescript
    responder({
      type: 'BUILD_ACTION',
      data: batchActions,
    } satisfies OffscreenResponse);
```

When you do, you may notice that there seems to be a typing issue?

Reply:
outdated, see updated flow